inc/class-wpseo-utils.php

```diff
@@ -483,35 +492,266 @@ public static function clear_transient_cache( $option ) {
 		}
+	}
 
-		$query = "DELETE FROM $wpdb->options WHERE";
+	/**
+	 * Invalidate sitemap cache
+	 *
+	 * @param null|string $type The type to get the key for. Null for all cache.
+	 */
+	private static function invalidate_sitemap_cache( $type = null ) {
+		// Global validator gets cleared when not type is provided.
+		$old_validator = null;
 
-		if ( ! empty( $types ) ) {
-			$query .= " option_name LIKE '_transient_wpseo_sitemap_cache_1_%' OR option_name LIKE '_transient_timeout_wpseo_sitemap_cache_1_%'";
+		// Get the current type validator.
+		if ( ! is_null( $type ) ) {
+			$old_validator = self::get_sitemap_cache_validator( $type );
+		}
 
-			foreach ( $types as $sitemap_type ) {
-				$query .= ' OR ';
-				$query .= " option_name LIKE '_transient_wpseo_sitemap_cache_" . $sitemap_type . "_%' OR option_name LIKE '_transient_timeout_wpseo_sitemap_cache_" . $sitemap_type . "_%'";
-			}
+		// Refresh validator.
+		self::new_sitemap_cache_validator( $type );
+
+		if ( ! wp_using_ext_object_cache() ) {
+			// Clean up current cache from the database.
+			self::cleanup_sitemap_database_cache( $type, $old_validator );
+		}
+
+		// External object cache pushes old and unretrieved items out by itself so we don't have to do anything for that.
+	}
+
+	/**
+	 * Cleanup invalidated database cache
+	 *
+	 * @param null|string $type The type of sitemap to clear cache for.
+	 * @param null|string $validator The validator to clear cache of.
+	 *
+	 * @return void
+	 */
+	private static function cleanup_sitemap_database_cache( $type = null, $validator = null ) {
+		global $wpdb;
+
+		if ( is_null( $type ) ) {
+			// Clear all cache if no type is provided.
+			$where = sprintf( "option_name LIKE '_transient_%s%%'", self::$sitemap_cache_key_prefix );
 		}
 		else {
-			$query .= " option_name LIKE '_transient_wpseo_sitemap_%' OR option_name LIKE '_transient_timeout_wpseo_sitemap_%'";
+			if ( ! is_null( $validator ) ) {
+				// Clear all cache for provided type-validator.
+				$where = sprintf( "option_name LIKE '_transient_%%_%s'", $validator );
+			}
+			else {
+				// Clear type cache for all type keys.
+				$where = sprintf( "option_name LIKE '_transient_%s%s_%%'", self::$sitemap_cache_key_prefix, $type );
+			}
 		}
 
+		$query = sprintf( 'DELETE FROM %s WHERE %s', $wpdb->options, $where );
 		$wpdb->query( $query );
 	}
 
 	/**
+	 * Get the sitemap cache key prefix
+	 *
+	 * @return string
+	 */
+	public static function get_sitemap_cache_key_prefix() {
+		return self::$sitemap_cache_key_prefix;
+	}
+
+	/**
+	 * Get the cache key for a certain type and page
+	 *
+	 * @param null|string $type The type to get the key for. Null or '1' for index cache.
+	 * @param int         $page The page of cache to get the key for.
+	 *
+	 * @return string The key where the cache is stored on.
+	 */
+	public static function get_sitemap_cache_key( $type = null, $page = 1 ) {
+		// Using '1' for sitemap index cache.
+		$type = is_null( $type ) ? '1' : $type;
```

Comment:
I would assign `'1'` to a class constant to be more descriptive. See this used before in filtering `$types`